src/main/cpp/src/hive_hash.cu

```diff
@@ -486,16 +526,70 @@ std::unique_ptr<cudf::column> hive_hash(cudf::table_view const& input,
 
   check_nested_depth(input);
 
+  // `flattened_column_views` only contains nested columns and columns that result from flattening
+  // nested columns
+  std::vector<cudf::column_view> flattened_column_views;
+  // `first_child_index` has the same size as `flattened_column_views`
+  std::vector<cudf::size_type> first_child_index;
+  // `nested_column_map` has the same size as `input.num_columns()`, and it maps the column index in
+  // `input` to the index in `flattened_column_views`
+  std::vector<cudf::size_type> nested_column_map;
+
+  for (auto i = 0; i < input.num_columns(); i++) {
+    auto const& root_col = input.column(i);
+    if (root_col.type().id() == cudf::type_id::LIST ||
+        root_col.type().id() == cudf::type_id::STRUCT) {
+      // Construct the `flattened_column_views` by level order traversal
+      nested_column_map.push_back(flattened_column_views.size());
+      flattened_column_views.push_back(root_col);
+      // flattened_column_views[idx] is the next column to process
+      for (auto idx = nested_column_map.back();
+           idx < static_cast<cudf::size_type>(flattened_column_views.size());
+           idx++) {
+        auto const col = flattened_column_views[idx];
```

Comment:
```suggestion
        auto const& col = flattened_column_views[idx];
```

Reply:
I've tried it before, but it produces many different types of bugs when running tests.😨

Reply:
This line should not cause a bug, so that means there is a bug hidden somewhere else. What are these bugs?

Reply:
```
[ERROR] Failures:
[ERROR]   HashTest.testHiveHashNestedStruct:739 Column unnamed Row 0 ==> expected: <89581538> but was: <89768068>
[ERROR]   HashTest.testHiveHashStruct:712 Column unnamed Row 0 ==> expected: <89581538> but was: <92471264>
[ERROR] Errors:
[ERROR]   HashTest.testHiveHashListOfStruct:843 ? Cudf CUDF failure at: /home/yafeng/spa...
[ERROR]   HashTest.testHiveHashLists:755 ? Cudf CUDF failure at: /home/yafeng/spark-rapi...
[ERROR]   HashTest.testHiveHashNestedLists:785 ? Cudf CUDF failure at: /home/yafeng/spar...
[ERROR]   HashTest.testHiveHashStructOfList:827 ? Cudf CUDF failure at: /home/yafeng/spa...
[INFO]
[ERROR] Tests run: 47, Failures: 2, Errors: 4, Skipped: 0
```
There are `Failures` and `Errors`, with two types of `Errors`.
```
[ERROR] testHiveHashStructOfList  Time elapsed: 0.014 s  <<< ERROR!
ai.rapids.cudf.CudfException: CUDF failure at: /home/yafeng/spark-rapids-jni/thirdparty/cudf/cpp/src/lists/lists_column_view.cu:36: lists column has an incorrect number of children
```
```
[ERROR] testHiveHashLists  Time elapsed: 0.001 s  <<< ERROR!
ai.rapids.cudf.CudfException: CUDF failure at: /home/yafeng/spark-rapids-jni/thirdparty/cudf/cpp/src/lists/lists_column_view.cu:29: lists_column_view only supports lists
```

Reply:
OK I was wrong. The line above indeed is problematic. It creates a reference to a vector item, but that reference is invalidated when the vector grows.